src/core/thread/mesh_forwarder_ftd.cpp

```diff
@@ -1020,6 +1023,144 @@ void MeshForwarder::UpdateRoutes(uint8_t *           aFrame,
+    {
+        VerifyOrExit((entry = FindFragmentPriorityEntry(aFragmentHeader.GetDatagramTag())) != NULL);
+
+        if (aFragmentHeader.GetDatagramOffset() + aFragmentLength >= aFragmentHeader.GetDatagramSize())
+        {
+            entry->SetLifetime(0);
+        }
+    }
+
+exit:
+    return;
+}
+
+FragmentPriorityEntry *MeshForwarder::FindFragmentPriorityEntry(uint16_t aTag)
+{
+    size_t i;
+
+    for (i = 0; i < OT_ARRAY_LENGTH(mFragmentEntries); i++)
+    {
+        if ((mFragmentEntries[i].GetLifetime() != 0) && (mFragmentEntries[i].GetDatagramTag() == aTag))
+        {
+            break;
+        }
+    }
+
+    return (i >= OT_ARRAY_LENGTH(mFragmentEntries)) ? NULL : &mFragmentEntries[i];
+}
+
+FragmentPriorityEntry *MeshForwarder::GetUnusedFragementPriorityEntry(void)
+{
+    size_t i;
+
+    for (i = 0; i < OT_ARRAY_LENGTH(mFragmentEntries); i++)
+    {
+        if (mFragmentEntries[i].GetLifetime() == 0)
+        {
+            break;
+        }
+    }
+
+    return (i >= OT_ARRAY_LENGTH(mFragmentEntries)) ? NULL : &mFragmentEntries[i];
+}
+
+otError MeshForwarder::GetFragmentPriority(Lowpan::FragmentHeader &aFragmentHeader, uint8_t &aPriority)
+{
+    otError                error = OT_ERROR_NONE;
+    FragmentPriorityEntry *entry;
+
+    VerifyOrExit((entry = FindFragmentPriorityEntry(aFragmentHeader.GetDatagramTag())) != NULL,
+                 error = OT_ERROR_NOT_FOUND);
+    aPriority = entry->GetPriority();
+
+exit:
+    return error;
+}
+
+otError MeshForwarder::GetForwardFramePriority(const uint8_t *     aFrame,
+                                               uint8_t             aFrameLength,
+                                               const Mac::Address &aMacDest,
+                                               const Mac::Address &aMacSource,
+                                               uint8_t &           aPriority)
+{
+    otError                error = OT_ERROR_NONE;
+    Lowpan::FragmentHeader fragmentHeader;
+
+    SuccessOrExit(error = SkipMeshHeader(aFrame, aFrameLength));
+
+    if (GetFragmentHeader(aFrame, aFrameLength, fragmentHeader) == OT_ERROR_NONE)
+    {
+        aFrame += fragmentHeader.GetHeaderLength();
+        aFrameLength -= fragmentHeader.GetHeaderLength();
+
+        if (fragmentHeader.GetDatagramOffset() == 0)
+        {
+            // Get priority from Ipv6 header or UDP destination port directly
+            SuccessOrExit(error = GetFramePriority(aFrame, aFrameLength, aMacSource, aMacDest, aPriority));
+        }
+        else
+        {
+            // Get priority from the pre-buffered info
+            SuccessOrExit(error = GetFragmentPriority(fragmentHeader, aPriority));
+        }
+
+        UpdateFragmentPriority(fragmentHeader, aFrameLength, aPriority);
+    }
+    else
+    {
+        // Get priority from Ipv6 header or UDP destination port directly
+        error = GetFramePriority(aFrame, aFrameLength, aMacSource, aMacDest, aPriority);
+    }
+
+exit:
+    return error;
```

Comment:
Add logs or accounting to report any errors?

Reply:
Changed.